library/asn1write.c

```diff
@@ -234,21 +234,33 @@ int mbedtls_asn1_write_bool( unsigned char **p, unsigned char *start, int boolea
 int mbedtls_asn1_write_int( unsigned char **p, unsigned char *start, int val )
 {
     int ret;
-    size_t len = 0;
+    size_t len;
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+    unsigned i;
+    unsigned char *c;
+#endif
 
-    if( *p - start < 1 )
+    if( *p < start || (size_t)( *p - start ) < sizeof(int) )
```

Comment:
We should not assume that the internal presentation of integers uses a two's complement here. Please test for the sign of `val` explicitly.